In [2]:
from sklearn import set_config

set_config(display="text")

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score
from sklearn.metrics import recall_score, f1_score, confusion_matrix, classification_report

In [4]:
data = load_breast_cancer()
print(type(data))
print(dir(data))

<class 'sklearn.utils._bunch.Bunch'>
['DESCR', 'data', 'data_module', 'feature_names', 'filename', 'frame', 'target', 'target_names']


In [5]:
x = data.data
# print(x)

y = data.target
print(x.shape)
print(len(y))


(569, 30)
569


In [6]:
print(data.DESCR)

.. _breast_cancer_dataset:

Breast cancer Wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

:Number of Instances: 569

:Number of Attributes: 30 numeric, predictive attributes and the class

:Attribute Information:
    - radius (mean of distances from center to points on the perimeter)
    - texture (standard deviation of gray-scale values)
    - perimeter
    - area
    - smoothness (local variation in radius lengths)
    - compactness (perimeter^2 / area - 1.0)
    - concavity (severity of concave portions of the contour)
    - concave points (number of concave portions of the contour)
    - symmetry
    - fractal dimension ("coastline approximation" - 1)

    The mean, standard error, and "worst" or largest (mean of the three
    worst/largest values) of these features were computed for each image,
    resulting in 30 features.  For instance, field 0 is Mean Radius, field
    10 is Radius SE, field 20 is Worst Radius.

    - 

In [7]:
# 데이터 분할(학습 80%, 테스트 20%)

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=42)
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)


(455, 30) (455,)
(114, 30) (114,)


In [8]:
# 모델 학습(로지스틱 회귀)

model = LogisticRegression(max_iter=10000, )
model.fit(x_train, y_train)
#print(dir(model))

LogisticRegression(max_iter=10000)

In [9]:
# 예측 (클래스 예측 및 확률 예측)
y_pred = model.predict(x_test)
print(y_pred)
print(y_pred.shape)
print(y_pred==y_test)
print(pd.Series(y_pred==y_test).unique())
print(pd.Series(y_pred==y_test).value_counts())

y_prob = model.predict_proba(x_test)
print(y_prob)

[1 0 0 1 1 0 0 0 1 1 1 0 1 0 1 0 1 1 1 0 1 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 0
 1 0 1 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 1 1 1 0 0 1 1 1 0 0 1 1 0 0 1 0
 1 1 1 1 1 1 0 1 1 0 0 0 0 0 1 1 1 1 1 1 1 1 0 0 1 0 0 1 0 0 1 1 1 0 1 1 0
 1 0 0]
(114,)
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True False  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True False  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True False  True  True  True  True False  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True False  True]
[ True False]
True     109
False      5
Name: count, dtype: int64

In [10]:
# 출력: 혼합행렬

cm = confusion_matrix(y_test, y_pred)
print(cm)

[[39  4]
 [ 1 70]]


In [11]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# 독립 실행을 위해 데이터와 모델을 직접 준비
X, y = load_breast_cancer(return_X_y=True)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LogisticRegression(max_iter=10000)
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

# 평가지표 계산
acc = accuracy_score(y_test, y_pred)  # 정확도
prec = precision_score(y_test, y_pred)  # 정밀도
rec = recall_score(y_test, y_pred)  # 재현율
f1 = f1_score(y_test, y_pred)  # F1 점수

print(acc)
print(prec)
print(rec)
print(f1)

0.956140350877193
0.9459459459459459
0.9859154929577465
0.9655172413793104


# 데이터가 불균형일때 발생하는 문제
불균형한 이진 분류로 만들기 위해 숫자 '7'만 True 나머지 숫자는 False로 설정

- 마찬가지로 fit은 아무것도 하지 않음
- predict에서는 무조건 0을 반환함
- 모든 예측을 다 0으로 하는 것! (모두 7이 아닌 다른 숫자라고 예측)
- 당연히 모델로서 말이 안 되는데 이것의 정확도를 측정

In [12]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.base import BaseEstimator
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score, f1_score
import numpy as np
import pandas as pd

In [13]:
class MyFakeClassifier(BaseEstimator):
    """항상 0을 예측하는 가짜 분류기"""
    def fit(self, X, y):
        pass

    def predict(self, X):
        return np.zeros(len(X), dtype=int)


# digits 데이터셋 로드
print("digits 데이터셋 로드 완료")
digit = load_digits()

digits 데이터셋 로드 완료


In [14]:
print(digit.target)

# 숫자가 7인 경우 True이고, 나머지는 False다.
# 이를 1과 0으로 변환하기 위해서 astype(int)로 한 것이다.
y = (digit.target == 7).astype(int)

print("\n라벨 분포:")
print(pd.Series(y).value_counts())

# 학습/테스트 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(digit.data, y, random_state=11)

print("\n훈련 세트 라벨 분포:")
print(pd.Series(y_train).value_counts())

print("\n테스트 세트 라벨 분포:")
print(pd.Series(y_test).value_counts())

[0 1 2 ... 8 9 8]

라벨 분포:
0    1618
1     179
Name: count, dtype: int64

훈련 세트 라벨 분포:
0    1213
1     134
Name: count, dtype: int64

테스트 세트 라벨 분포:
0    405
1     45
Name: count, dtype: int64


In [15]:
fakeclf = MyFakeClassifier()
fakeclf.fit(X_train, y_train)

fakpred = fakeclf.predict(X_test)
print('모든 예측을 0으로 하여도 정확도는 {:.3f}'.format(accuracy_score(y_test, fakpred)))

cf = confusion_matrix(y_test, fakpred)
print('\n혼동 행렬:')
print(cf)

# F1 score는 이진 분류에서 라벨이 0/1 형태일 때 계산 가능
print('\nF1 score:', f1_score(y_test, fakpred))

모든 예측을 0으로 하여도 정확도는 0.900

혼동 행렬:
[[405   0]
 [ 45   0]]

F1 score: 0.0


In [16]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.base import BaseEstimator
from sklearn.metrics import confusion_matrix, accuracy_score, f1_score
import numpy as np


class MyFakeClassifier(BaseEstimator):
    def fit(self, X, y):
        pass

    def predict(self, X):
        return np.zeros(len(X), dtype=int)


digit = load_digits()
y = (digit.target == 7).astype(int)

X_train, X_test, y_train, y_test = train_test_split(digit.data, y, random_state=11)

fakeclf = MyFakeClassifier()
fakeclf.fit(X_train, y_train)

fakpred = fakeclf.predict(X_test)

cf = confusion_matrix(y_test, fakpred)
print('혼동 행렬:')
print(cf)

f1 = f1_score(y_test, fakpred)
print('\nF1 score:', f1)
print('Accuracy:', accuracy_score(y_test, fakpred))

혼동 행렬:
[[405   0]
 [ 45   0]]

F1 score: 0.0
Accuracy: 0.9


- 모든 예측을 0으로 했지만, 0 (7이 아닌 숫자)이 1 (숫자 7)보다 훨씬 많은 불균형 데이터였기 때문에, 정확도가 무려 90%가 나왔다.

- 이렇게 두 가지 예제로 불균형 데이터에서 정확도의 성능 평가지표로서 한계를 보였다

#### average: Literal['micro', 'macro', 'samples', 'weighted', 'binary'] | None = 
#### 1. average=None


0. average='binary' (default)
   - positive label을 1로 간주하고 계산한다.
   - y_true에 0과 1 두 개 클래스만 있어야 한다.

1. average=None (클래스별 성능 개별 출력)

| 클래스               | Precision | Recall | F1-score | Support |
| ----------------- | --------- | ------ | -------- | ------- |
| **0 (malignant)** | 0.97      | 0.91   | 0.94     | 43      |
| **1 (benign)**    | 0.95      | 0.99   | 0.97     | 71      |

- 각 클래스별로 정밀도, 재현율, F1-score를 개별적으로 계산하여 보여준다.
- 모델이 클래스 0 (악성 종양)을 상대적으로 덜 잘 맞추고 있다는 것을 파악할 수 있다.


2. average='micro' (전체 기준 - 샘플 단위의 합)  
- micro 평균은 모든 TP, FP, FN을 전 클래스 기준으로 합산한 뒤 계산한다.
- 이 경우 정확도(Accuracy) = Precision = Recall = F1-score가 된다.
- 총 정답 수: TP(전체 맞춤 수) = 0 클래스에서 39개 + 1 클래스에서 70개 = 109
- 전체 샘플 수: 114
- 따라서:    Accuracy = Precision = Recall = F1 = 109 / 114 ≈ 0.9561  

3. average='macro' (단순 평균)  
- 클래스별 점수를 동일한 비중으로 평균함
  Macro Precision = (0.97 + 0.95) / 2 = 0.96  
  Macro Recall    = (0.91 + 0.99) / 2 = 0.95  
  Macro F1-score  = (0.94 + 0.97) / 2 = 0.955 ≈ 0.95  

- 클래스 0과 1의 비중이 같다고 가정하고 평균 내기 때문에, 클래스 불균형이 있는 경우에는 왜곡이 있을 수 있다.

4. average='weighted' (샘플 수 기반 가중 평균)  
-  각 클래스의 점수에 해당 클래스의 샘플 개수(Support)를 곱해서 평균
  Weighted Precision = (0.97×43 + 0.95×71) / 114 ≈ 0.958    
  Weighted Recall    = (0.91×43 + 0.99×71) / 114 ≈ 0.961  
  Weighted F1-score  = (0.94×43 + 0.97×71) / 114 ≈ 0.961  

- 실제 출력값은 모두 반올림되어 0.96으로 나옴

- 데이터 불균형을 고려한 평균이므로, 실전에서 가장 신뢰도 있는 지표 중 하나임

5. average 옵션 요약

| 평균 방식      | 설명                    | 특징               | 언제 사용하면 좋은가?                 |
| ---------- | --------------------- |------------ | ---------------------------- |
| `None`     | 클래스별 개별 지표 확인         |  `array` 형태       | 특정 클래스 성능 비교 시               |
| `micro`    | 전체 샘플 기준으로 정밀도/재현율 계산 | 불균형 데이터에 취약      | 다중 클래스, 불균형 무시하고 전체 성능 확인 시  |
| `macro`    | 클래스별 단순 평균            | 클래스 중요도 균일할 때 사용 |클래스 간 균형 평가 시                |
| `weighted` | 클래스별 가중 평균            | 불균형 데이터에 적합      |**불균형 데이터에서의 전체 성능 평가** 시 유리 |



In [17]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, recall_score, f1_score

# 독립 실행을 위해 데이터와 모델을 직접 준비
X, y = load_breast_cancer(return_X_y=True)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LogisticRegression(max_iter=10000)
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

print("average='micro'")
print("Precision:", precision_score(y_test, y_pred, average='micro'))
print("Recall   :", recall_score(y_test, y_pred, average='micro'))
print("F1 Score :", f1_score(y_test, y_pred, average='micro'))
print()

print("average='macro'")
print("Precision:", precision_score(y_test, y_pred, average='macro'))
print("Recall   :", recall_score(y_test, y_pred, average='macro'))
print("F1 Score :", f1_score(y_test, y_pred, average='macro'))
print()

print("average='weighted'")
print("Precision:", precision_score(y_test, y_pred, average='weighted'))
print("Recall   :", recall_score(y_test, y_pred, average='weighted'))
print("F1 Score :", f1_score(y_test, y_pred, average='weighted'))
print()

average='micro'
Precision: 0.956140350877193
Recall   : 0.956140350877193
F1 Score : 0.956140350877193

average='macro'
Precision: 0.960472972972973
Recall   : 0.9464461185718964
F1 Score : 0.9526381387619444

average='weighted'
Precision: 0.9569049312470365
Recall   : 0.956140350877193
F1 Score : 0.9558014271241044



In [18]:
from sklearn.metrics import classification_report


target_names = data.target_names  # ['malignant', 'benign']
print(target_names)

'''
| 용어            | 한글 해석   | 설명                                                                          |
| -------------   | ------- | --------------------------------------------------------------------------- |
| **malignant**   | 악성 (암)  | 빠르게 자라고 전이할 가능성이 높은 종양이나 질병. 보통 **암(Cancer)** 을 뜻함. 치료가 어렵고 생명에 위협이 될 수 있음. |
| **benign**      | 양성  (비암성) | 성장하긴 하지만 전이하지 않으며 비교적 **위험하지 않은 상태**. 수술 등으로 완치 가능한 경우가 많음.                 |

- malignant: 악성 종양 (암) → 0 또는 Negative Class
- benign: 양성 종양 (비암성) → 1 또는 Positive Class
'''



# 0. 전체 classification_report
print("Classification Report")
# target_names옵션을 설정하지 않으면 0과 1로 표시된다.
print(classification_report(y_test, y_pred))
print("Classification Report")
print(classification_report(y_test, y_pred, target_names=target_names))

['malignant' 'benign']
Classification Report
              precision    recall  f1-score   support

           0       0.97      0.91      0.94        43
           1       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114

Classification Report
              precision    recall  f1-score   support

   malignant       0.97      0.91      0.94        43
      benign       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



# 1. LogisticRegression  분류 알고리즘

LogisticRegression( )알고리즘의 penalty='l2', *, dual=False, tol=0.0001, C=1.0, fit_intercept=True, intercept_scaling=1, class_weight=None, random_state=None, solver='lbfgs', max_iter=100, multi_class='deprecated', verbose=0, warm_start=False, n_jobs=None, l1_ratio=None  


1. **penalty:**  
정규화 방법을 지정한다. 'l2'는 L2 정규화를 의미하며, 이는 가중치의 제곱합을 최소화하여 과적합을 방지하는 데 도움을 준다. 'l1'을 선택하면 L1 정규화를 사용하여 일부 가중치를 0으로 만들 수 있다. 'elasticnet'을 선택하면 L1과 L2 정규화를 혼합하여 사용할 수 있다.  
**2. dual:**  
이진 분류 문제에서 이중 문제를 해결할지를 결정한다. True로 설정하면 이중 문제를 해결하고, False로 설정하면 원래의 문제를 해결한다. 일반적으로 이진 분류에서는 False로 설정하는 것이 더 효율적이다.  
**3. tol:**  
수렴 기준을 설정합니다. 알고리즘이 수렴했다고 판단하는 기준으로, 기본값은 0.0001이다. 이 값이 작을수록 더 정밀한 수렴을 요구한다.  
**4. C:**  
정규화 강도를 조절하는 하이퍼파라미터이다. 값이 작을수록 정규화가 강해지고, 값이 클수록 정규화가 약해진다. 기본값은 1.0이다.  
**5. fit_intercept:**  
절편(intercept)을 모델에 포함할지를 결정한다. True로 설정하면 절편을 포함하고, False로 설정하면 절편을 포함하지 않는다. 기본값은 True이다.  
**6. intercept_scaling:**  
다중 클래스 문제에서 절편을 조정하는 데 사용된다. 기본값은 1이며, 이 값은 절편의 스케일을 조정하는 데 사용된다.  
**7. class_weight:**  
클래스의 가중치를 설정한다. None으로 설정하면 모든 클래스의 가중치가 동일하게 설정됩니다. 클래스 불균형이 있는 경우, 특정 클래스에 더 높은 가중치를 부여하여 모델이 해당 클래스를 더 잘 학습하도록 할 수 있다.
**8. random_state:**  
난수 생성기의 시드를 설정합니다. 동일한 시드를 사용하면 동일한 결과를 재현할 수 있다. 기본값은 None이다.  
**9. solver:**  
최적화 알고리즘을 선택한다. 'lbfgs', 'newton-cg', 'liblinear', 'sag', 'saga' 중 하나를 선택할 수 있다. 각 솔버는 서로 다른 특성과 성능을 가지므로 데이터셋에 따라 적절한 솔버를 선택하는 것이 중요하다.  
**10. max_iter:**      
알고리즘의 최대 반복 횟수를 설정한다. 기본값은 100이며, 이 값에 도달하기 전에 수렴하면 알고리즘이 종료된다.  
**11. multi_class:**  
다중 클래스 문제에서의 처리 방법을 설정한다. 'deprecated'는 더 이상 사용되지 않는 옵션이며, 대신 'ovr' (One-vs-Rest) 또는 'multinomial'을 사용할 수 있다.  
**12. verbose:**  
출력의 상세 정도를 설정한다. 0이면 출력이 없고, 1 이상의 값으로 설정하면 더 많은 정보가 출력된다.  
**13. warm_start:**  
이전에 학습한 모델을 재사용할지를 결정한다. True로 설정하면 이전 모델의 가중치를 초기값으로 사용하여 학습을 계속할 수 있다.  
**14. n_jobs:**  
병렬 처리를 위한 작업 수를 설정한다. None으로 설정하면 기본값이 사용되며, -1로 설정하면 모든 CPU 코어를 사용한다.  
**15. l1_ratio:**  
L1과 L2 정규화의 비율을 설정한다. 'elasticnet'을 사용할 때만 의미가 있으며, 기본값은 None이다.

### 타이타닉 탑승자 데이터
| 컬럼명              | 설명                                                        |
| ---------------- | --------------------------------------------------------- |
| **survived**     | 생존 여부 (0 = 사망, 1 = 생존)                                    |
| **pclass**       | 선실 등급 (1 = 1등석, 2 = 2등석, 3 = 3등석)                         |
| **sex**          | 성별 ('male', 'female')                                     |
| **age**          | 나이                                                        |
| **sibsp**        | 함께 탑승한 형제/배우자 수 (siblings/spouses)                        |
| **parch**        | 함께 탑승한 부모/자녀 수 (parents/children)                         |
| **fare**         | 승객 요금                                                     |
| **embarked**     | 탑승 항구 코드 (C = Cherbourg, Q = Queenstown, S = Southampton) |
| **class**        | `pclass`와 동일하지만 문자열로 표시 ('First', 'Second', 'Third')      |
| **who**          | 승객 구분 ('man', 'woman', 'child')                           |
| **adult\_male**  | 성인 남성 여부 (True/False)                                     |
| **deck**         | 객실이 위치한 갑판 (A\~G 등, 일부 결측값 존재)                            |
| **embark\_town** | 탑승 도시명 (예: 'Cherbourg', 'Southampton')                    |
| **alive**        | 생존 여부 문자열 ('yes' 또는 'no')                                 |
| **alone**        | 혼자 탑승 여부 (True/False)                                     |

※ survived와 alive은 같은 컬럼이므로 alive제외하고 훈련해야함


In [19]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score,confusion_matrix,  classification_report

In [20]:
titanic = sns.load_dataset('titanic')
print(titanic.columns)
print(titanic.info())
print(titanic.head(2))


Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        

In [21]:
titanic.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [22]:
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [23]:
titanic.fillna({'age': titanic['age'].mean()}, inplace=True)

In [24]:
titanic['age'].isnull().sum()

np.int64(0)

In [25]:
titanic['embarked'].value_counts()

embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [26]:
# titanic['deck'] = titanic['deck'].cat.add_categories('N') # 일회성 
titanic.fillna({'deck': 'N'}, inplace=True)
titanic['deck'].isnull().sum()

TypeError: Cannot setitem on a Categorical with a new category (N), set the categories first

In [ ]:
titanic['deck'].value_counts()

deck
N    688
C     59
B     47
D     33
E     32
A     15
F     13
G      4
Name: count, dtype: int64

In [ ]:
pd.__version__

'2.3.3'

In [ ]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          891 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         891 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [ ]:
# 남, 여 탑승 수
titanic['sex'].value_counts()

sex
male      577
female    314
Name: count, dtype: int64

In [ ]:
# 생존을 0(사망): 548, 1(생존):342
titanic['survived'].value_counts()

survived
0    549
1    342
Name: count, dtype: int64

In [ ]:
# 성별 생존여부
# titanic.groupby(['sex'])['survived'].value_counts()
titanic.groupby(['sex', 'survived']).value_counts()

: 

In [ ]:
# 선실등급별 남여 생존여부
sns.barplot(data=titanic, x='sex', y='survived')


# 객실 등급별 성별에 따른 생존 확률

- 여성의 경우 일등실, 이등실에 따른 생존 확률의 차이는 크지 않으나, 삼등실의 경우 생존 확률이 상대적으로 많이 떨어짐을 알 수 있다.
- 남성의 경우는 일등실의 생존 확률이 이, 삼등실의 생존 확률보다 월등이 높다.
- 제임스 카메론의 영화인 '타이타닉'에서도 나왔듯이 바다에서 사고가 날 경우 여성과 아이들, 그리고 노약자가 제일 먼저 구조대상이다. 그리고 부자나 유명인이 다음 구조 대상이다. 삼등실에 탄 많은 가난한 이는 타이타닉 호와 운명을 함께 했다.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(titanic.loc[:, 'pclass': 'alone'], titanic['survived'], random_state=11)


In [ ]:
import platform 

if platform.system() == 'Windows':
    

model = LogisticRegression